src/framerate_gui.cpp

```diff
@@ -0,0 +1,752 @@
+		SetDParam(0, tpl);
+		SetDParam(1, (int)(value * 100));
+		SetDParam(2, 2);
+	}
+
+	virtual void SetStringParameters(int widget) const
+	{
+		double value;
+
+		switch (widget) {
+			case WID_FRW_CAPTION:
+				if (!this->small) break;
+				value = _pf_data[PFE_GAMELOOP].GetRate();
+				SetDParamGoodWarnBadRate(value, PFE_GAMELOOP);
+				value /= _pf_data[PFE_GAMELOOP].expected_rate;
+				SetDParam(3, (int)(value * 100));
+				SetDParam(4, 2);
+				break;
+
+			case WID_FRW_RATE_GAMELOOP:
+				value = _pf_data[PFE_GAMELOOP].GetRate();
+				SetDParamGoodWarnBadRate(value, PFE_GAMELOOP);
+				break;
+			case WID_FRW_RATE_DRAWING:
+				value = _pf_data[PFE_DRAWING].GetRate();
+				SetDParamGoodWarnBadRate(value, PFE_DRAWING);
+				break;
+			case WID_FRW_RATE_FACTOR:
+				value = _pf_data[PFE_GAMELOOP].GetRate();
+				value /= _pf_data[PFE_GAMELOOP].expected_rate;
+				SetDParam(0, (int)(value * 100));
+				SetDParam(1, 2);
+				break;
+			case WID_FRW_INFO_DATA_POINTS:
+				SetDParam(0, NUM_FRAMERATE_POINTS);
+				break;
+		}
+	}
+
+	virtual void UpdateWidgetSize(int widget, Dimension *size, const Dimension &padding, Dimension *fill, Dimension *resize)
+	{
+		switch (widget) {
+			case WID_FRW_RATE_GAMELOOP:
+				SetDParamGoodWarnBadRate(9999.99, PFE_GAMELOOP);
+				*size = GetStringBoundingBox(STR_FRAMERATE_RATE_GAMELOOP);
+				break;
+			case WID_FRW_RATE_DRAWING:
+				SetDParamGoodWarnBadRate(9999.99, PFE_DRAWING);
+				*size = GetStringBoundingBox(STR_FRAMERATE_RATE_BLITTER);
+				break;
+			case WID_FRW_RATE_FACTOR:
+				SetDParam(0, 999999);
+				SetDParam(1, 2);
+				*size = GetStringBoundingBox(STR_FRAMERATE_SPEED_FACTOR);
+				break;
+
+			case WID_FRW_TIMES_NAMES: {
+				int linecount = PFE_MAX - PFE_FIRST;
+				size->width = 0;
+				size->height = FONT_HEIGHT_NORMAL * (linecount + 1) + VSPACING;
+				for (int line = 0; line < linecount; line++) {
+					auto line_size = GetStringBoundingBox(STR_FRAMERATE_GAMELOOP + line);
+					size->width = max(size->width, line_size.width);
+				}
+				break;
+			}
+
+			case WID_FRW_TIMES_CURRENT:
+			case WID_FRW_TIMES_AVERAGE: {
+				int linecount = PFE_MAX - PFE_FIRST;
+				*size = GetStringBoundingBox(STR_FRAMERATE_CURRENT + (widget - WID_FRW_TIMES_CURRENT));
+				SetDParamGoodWarnBadDuration(9999.99);
+				auto item_size = GetStringBoundingBox(STR_FRAMERATE_VALUE);
+				size->width = max(size->width, item_size.width);
+				size->height += FONT_HEIGHT_NORMAL * linecount + VSPACING;
+				break;
+			}
+
+			default:
+				Window::UpdateWidgetSize(widget, size, padding, fill, resize);
+				break;
+		}
+	}
+
+	typedef double (ElementValueExtractor)(PerformanceElement elem);
+	void DrawElementTimesColumn(const Rect &r, int heading_str, ElementValueExtractor get_value_func) const
+	{
+		int y = r.top;
+		DrawString(r.left, r.right, y, heading_str, TC_FROMSTRING, SA_CENTER);
+		y += FONT_HEIGHT_NORMAL + VSPACING;
+
+		for (auto e = PFE_FIRST; e < PFE_MAX; e = (PerformanceElement)(e + 1)) {
```

Comment:
Cast is odd, e++ should suffice?

Reply:
MSVC 2015 complains about ++ operator not being valid for the type.

> error C2678: binary '++': no operator found which takes a left-hand operand of type 'PerformanceElement' (or there is no acceptable conversion)

I do see that some other enum types have had a custom `operator++` implemented for them.

Reply:
https://github.com/OpenTTD/OpenTTD/blob/master/src/core/enum_type.hpp#L16 ?